### PR TITLE
feat: support replies

### PR DESCRIPTION
**Pull request by thisislawatts**
PR Checklist

- [x] - Commit [messages are descriptive](https://github.com/ONEARMY/community-platform/blob/master/CONTRIBUTING.md#--commit-style-guide), it will be used in our [Release Notes](https://github.com/ONEARMY/community-platform/releases/)

PR Type

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Developer experience (improves developer workflows for contributing to the project)

## Description

* Adds the ability to reply to comments on Question items.

Screenshot 2024-01-31 at 18 23 26

**Note:** Design detailing will be addressed in subsequent PRs, the focus of this PR is the functional change.

### CODE REVIEW

File: packages/components/package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "dev": "tsc --watch",
     "lint": "eslint . --ext .js,.jsx,.ts,.tsx src --color",
     "new-component": "ts-node scripts/newComponent.ts",
-    "test": "vitest --coverage"
+    "test": "vitest"
```
Comment (thisislawatts) [inline packages/components/package.json:17]: Toggled this off for iterating, probably worth only running coverage in CI context for reporting. What do folks think?

Comment (thisislawatts) [reply packages/components/package.json:17]: Split out to a standalone item here:
https://github.com/ONEARMY/community-platform/pull/3240

File: src/common/transFormToUserComments.ts
Comment (benfurber) [inline src/common/transFormToUserComments.ts:1]: Nitpick? Transform is one word so the 'F' in the file name should be lowercase.

File: src/pages/Question/QuestionComments.tsx [hunk 2]
```diff
@@ -84,17 +86,22 @@ export const QuestionComments = ({
...
+              discussionObject,
+              reply,
+              commentId,
+            )
+            commentsUpdated &&
+              commentsUpdated(
+                transformToUserComments(updatedObj?.comments || [], activeUser),
+              )
+          }
+        }}
```
Comment (benfurber) [inline src/pages/Question/QuestionComments.tsx:102]: Pull this into a function above?

**Review by benfurber — APPROVED** [review]
Looking great!

File: src/common/transFormToUserComments.ts
Comment (thisislawatts) [reply src/common/transFormToUserComments.ts:1]: Good catch, updated.

**Comment by thisislawatts** [comment]
Updated @davehakkens

File: packages/components/src/DiscussionContainer/DiscussionContainer.tsx [hunk 3]
```diff
@@ -25,25 +29,76 @@ export const DiscussionContainer = (props: IProps) => {
...
     onMoreComments,
     onSubmit,
     isLoggedIn,
+    supportReplies: _supportReplies,
   } = props
 
+  const [commentBeingRepliedTo, setCommentBeingRepliedTo] = useState<
+    null | string
+  >(null)
+  const supportReplies = _supportReplies ?? false
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply packages/components/src/DiscussionContainer/DiscussionContainer.tsx]: Updated here, stumbled on a neat pattern for defining defaults  7c3f98b7057d73b2a3b7b3310f27de36dd12d44f

File: packages/components/src/DiscussionContainer/DiscussionContainer.tsx [hunk 4]
```diff
@@ -25,25 +29,76 @@ export const DiscussionContainer = (props: IProps) => {
...
+                    onSubmitReply(commentId, comment)
+                  }
+                  setCommentBeingRepliedTo(null)
+                }}
+                buttonLabel="Send your reply"
+                isLoggedIn={isLoggedIn}
+              />
+            </Box>
+          )
+        }}
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply packages/components/src/DiscussionContainer/DiscussionContainer.tsx]: Moved to a separate function definition

File: packages/components/src/DiscussionContainer/DiscussionContainer.tsx [hunk 5]
```diff
@@ -25,25 +29,76 @@ export const DiscussionContainer = (props: IProps) => {
...
+        comments={structuredComments}
         handleDelete={handleDelete}
         handleEdit={handleEdit}
         handleEditRequest={handleEditRequest}
         highlightedCommentId={highlightedCommentId}
         onMoreComments={onMoreComments}
+        setCommentBeingRepliedTo={(commentId) => {
```
[the remark this answers is not in the document]
Comment (thisislawatts) [reply packages/components/src/DiscussionContainer/DiscussionContainer.tsx]: Updated here: https://github.com/ONEARMY/community-platform/pull/3225/commits/d5ea1b1ce70f76595351c6c8c82cb3827142b94e

**Pull request merged**